### DISCUSSION TITLE
Character list

### DISCUSSION THREAD

**Discussion by iandoug** [binarybottle/engram-es#13] [General]
Hi

Can someone post the desired character list?

Thanks, Ian

**Comment by Lobo-Feroz** [comment]
Hey Ian, let me know if this is what you're looking for:

Characters: 
```
abcdefghijklmnopqrstuvwxyz 
ABCDEFGHIJKLMNOPQRSTUVWXYZ
```

Tilded characters:
```
áéíóúüñç
ÁÉÍÓÚÜÑÇ
```
(the ç is only used in the catalan language but I think it's not too much of a hassle to keep it as an altgr version of the c)

Spanish Punctuation Symbols:
`<>'*|¶ ,«»“[]¨:-_{}()§.;— ¿?¡!´`

In spanish opening and closing quotation marks (") are different, but here in github they look the same. Mostly world processors decorate them on the fly. Same for single qm's, here they look like two apostrophes, so I haven't listed them.

Other symbols:
```
ºª@#·~$€%&¬\/^`+=
```   

There might be also other characters and symbols that are not on the qwerty-es keyboard that we could consider including for their usefulness.
For instance, maybe the pound £ and the yen ¥

@NickG13 do you think there are any missing character or symbol?

**Comment by iandoug** [comment]
"In spanish opening and closing quotation marks (") are different, but here in github they look the same. Mostly world processors decorate them on the fly. Same for single qm's, here they look like two apostrophes, so I haven't listed them."

So you have ASCII quotes ' and " (U+0027 U+0022) as well as typographic quotes ‘’“” (2018, 2019, 201C, 201D) ?

Are ¨ and ´ dead characters just for diacritics?

Regarding moving punctuation around, Shai wrote about that: 
https://colemak.com/Design
See Punctuation section.

The gender(?) indicators ºª seem to be seldom used?

Thanks, Ian

**Reply by Lobo-Feroz** [reply]
> So you have ASCII quotes ' and " (U+0027 U+0022) as well as typographic quotes ‘’“” (2018, 2019, 201C, 201D) ?

Yes, that's correct! The current situation is that the qwerty-es keyboard only has the ASCII quotes, and some text processors like MS Word render them as the correct typographic quotes. I don't know how this is solved in other languages.

The single ASCII quote ' , if it's the same as the apostrophe, is very seldom used on its own in spanish. We can have contractions like "it's" or "don't" but it's not proper spanish language. And of course, it's used when quoting some english text, like the name of a place or a movie. So it's needed but way, way less common than in english.

> Are ¨ and ´ dead characters just for diacritics?

You mean if we never write them on their own? Yeah that'd be right. No use for standalone ¨ or ´ (but we can write them with our keyboards, just hit the key then space). Same for ` and ^.

> The gender(?) indicators ºª seem to be seldom used?

Yes, they are for ordinals. For instance to say you live in the first (1st) floor, you could say you live in the "primero", or the "1º".

I wouldn't say that they are too rare without the data. For instance you use them when writing your address. It's one of those things that I'd like to see how much they are actually used.

> Regarding moving punctuation around, Shai wrote about that:
https://colemak.com/Design
See Punctuation section.

Thanks for the info! Will read it thoroughly.

**Reply by iandoug** [reply]
> Yes, that's correct! The current situation is that the qwerty-es keyboard only has the ASCII quotes, and some text processors like MS Word render them as the correct typographic quotes. I don't know how this is solved in other languages.

The same way.. (well, for English. The new German layout includes them. It also makes coffee.
https://en.wikipedia.org/wiki/QWERTZ#/media/File:German-Keyboard-Layout-T2-Version1-large.png )

The EurKeys layout includes them, on AltGr
https://eurkey.steffen.bruentjen.eu/layout.html

I think typographic quotes would need to go on AltGr. On my current layout, I have [{ ]} on the top row and single and double typographic quotes on the AltGr slots on them. But don't need them very often.

Cheers, Ian

**Reply by NickG13** [reply]
Hi @iandoug , its a pleasure having you here (I'm EnergyRaising on Reddit) As far as I know, in spanish we can use quotes ' when they are done inside " quotes, as a second resource, I don't know if that's what Lobo-Feroz is saying. 

As per Shai instructions, what I don't like is the fact of trying to adjust the punctuation to stetics and mobile phones. We aren't building a keyboard layout to be popular, pretty or usable on phone (a lot of users said some layouts aren't needed neither compatible with phones), as I understand, we are trying to build here the most optimized and comfortable layout possible for spanish. The logic behind punctuation placement also feels important to me, though, and I think Arno did it very well with the english version. MAybe we can follow the same principles.

As per AltGr, well, depending on the punctuation you put there, you can make the typing experience a lot slower and uncomfortable. Also, if you are using ANSI keyboards you should press Ctrl + alt, which makes it even worse. I want to remember that there's always the option on doing some mods (for example, like Colemak Wide versions), which make AltGr more ergonomical for the thumb.

**Reply by iandoug** [reply]
> Also, if you are using ANSI keyboards you should press Ctrl + alt, which makes it even worse

I think that only applies to keyboards set to US English, in the OS. I have original MS Natural and have no issues using AltGr on it directly. Am on Linux.

**Comment by iandoug** [comment]
Big screen version. I dropped the typographic quotes for now, even though they are a nice-to-have, can probably find suitable spots for them once more important characters are sorted.
Your list seems to have both n-dash and m-dash, but I can't find m-dash on Wiki's pic of the Spanish Qwerty.

What's the feeling regarding ZXCV? 
In recent months I've been testing my own layouts, one of which did not have ZXCV, and found it a real annoyance. 

Anyway, pic attached for discussion ... I grouped them a bit differently.
![spanish-chars](https://user-images.githubusercontent.com/12673886/127750664-27d7290a-9f88-4761-b156-2b64802afad9.png)

**Reply by iandoug** [reply]
I need this list to be correct in order to clean up the corpus files to compute character and n-gram frequencies.
Anything not on here gets fixed or deleted.

**Reply by binarybottle** [reply]
> What's the feeling regarding ZXCV?
> In recent months I've been testing my own layouts, one of which did not have ZXCV, and found it a real annoyance.

I have it set up on my Mac so that if I press the command key then the keys revert to qwerty, so ZXCV is retained.

**Reply by Lobo-Feroz** [reply]
> I need this list to be correct in order to clean up the corpus files to compute character and n-gram frequencies.
Anything not on here gets fixed or deleted.

I think ~ can appear as a dead key (~ + N = Ñ) but also on its own meaning "more or less", just as in english. E.g. this keyboard will have ~124 characters.

The rest seem ok to me. You won't see much of § or ¶ btw, probably close to zero.

I can't really say if there will be other characters in the corpus. Meaning, if I've missed something, I won't notice it looking at the list. ~Do you have any text file that I could check?~ Will be checking the Leipzig corpora just in case.

**Reply by iandoug** [reply]
On English Qwerty, ~ ` ^ are characters like any other. On Linux, when you use the Compose key to do things like é or ó, then we use the apostrophe ' . I can also use ~ ` ^ in the same way, ñ à or ê. For the diaresis, we use the double quote ", ë. But you guys have dedicated ¨ and  ´ keys.

I've just downloaded the Leipzig files, will see what's inside to tomorrow. Getting late :-)

**Comment by iandoug** [comment]
Okay, I have a 3.4GB file, but there are a ton of Chinese characters, here is one place where they show up:
`697967	No me gusta Apple en general su iphone, sistema operativo, etc. pero admito que cuando vi esto me sorprendió.
697968	No me gusta Apple en general su iphone, sistema operativo, etc. pero admito que cuando vi esto me sorprendi贸.`

So I'm guessing there is a bug in UniLeipzig's extraction process, since 
1. The sentence is duplicated, and 
2. Chinese should be ó.

I think the simplest 'cleanup' procedure will be to simply discard any sentences with characters that are NOT in the Spanish Keyboard list.

**Reply by Lobo-Feroz** [reply]
I can confirm that, that chinese character in that context is clearly and unequivocally supposed to be an ó.
